src/i2c.rs

```diff
@@ -249,37 +264,157 @@ where
+
+        if is_write {
+            // Send a START condition
+            self.i2c.cr1.modify(|_, w| w.start().set_bit());
+        } else {
+            // Send a START condition
+            // Set ACK bit for read operations
+            self.i2c
+                .cr1
+                .modify(|_, w| w.start().set_bit().ack().set_bit());
+        }
+
+        // Wait until START condition was generated
+        while self.check_and_clear_error_flags()?.sb().bit_is_clear() {}
+
+        // Also wait until signalled we're master and everything is waiting for us
+        loop {
+            self.check_and_clear_error_flags()?;
+
+            let sr2 = self.i2c.sr2.read();
+            if !(sr2.msl().bit_is_clear() && sr2.busy().bit_is_clear()) {
+                break;
+            }
+        }
+
+        // Send out address
+        self.i2c.dr.write(|w| unsafe { w.bits(wr_addr) });
+
+        // Wait until address was sent
+        while self.check_and_clear_error_flags()?.addr().bit_is_clear() {}
+
+        // Clear condition by reading SR2
+        self.i2c.sr2.read();
+
+        Ok(())
+    }
+
+    fn send_byte(&self, byte: u8) -> Result<(), Error> {
+        // Wait until we're ready for sending
+        // Check for any I2C errors. If a NACK occurs, the ADDR bit will never be set.
+        while self.check_and_clear_error_flags()?.txe().bit_is_clear() {}
+
         // Push out a byte of data
         self.i2c.dr.write(|w| unsafe { w.bits(u32::from(byte)) });
 
-        self.check_and_clear_error_flags(&self.i2c.sr1.read())?;
+        // Wait until byte is transferred
+        // Check for any potential error conditions.
+        while self.check_and_clear_error_flags()?.btf().bit_is_clear() {}
+
         Ok(())
     }
 
     fn recv_byte(&self) -> Result<u8, Error> {
-        loop {
-            let sr = self.i2c.sr1.read();
-            self.check_and_clear_error_flags(&sr)?;
-            if sr.rxne().bit_is_set() {
-                break;
+        while self.check_and_clear_error_flags()?.rxne().bit_is_clear() {}
+        let value = self.i2c.dr.read().bits() as u8;
+        Ok(value)
+    }
+
+    fn read_bytes(&mut self, buffer: &mut [u8]) -> Result<(), Error> {
+        // Receive bytes into buffer
+        for c in buffer {
+            *c = self.recv_byte()?;
+        }
+
+        Ok(())
+    }
+    /// Reads like normal but does'n generate start and don't send address
+    fn read_wo_prepare(&mut self, buffer: &mut [u8]) -> Result<(), Error> {
+        if let Some((last, buffer)) = buffer.split_last_mut() {
+            // Read all bytes but not last
+            self.read_bytes(buffer)?;
+
+            // Prepare to send NACK then STOP after next byte
+            self.i2c
+                .cr1
+                .modify(|_, w| w.ack().clear_bit().stop().set_bit());
+
+            // Receive last byte
+            *last = self.recv_byte()?;
+
+            // Wait for the STOP to be sent. Otherwise, the interface will still be
```

Comment:
The comment refers to nb:Error and DMA, neither of which are in use in this i2c module, should be removed